torch/utils/data/datapipes/iter/combinatorics.py

```diff
@@ -166,13 +173,16 @@ def __setstate__(self, state):
             self.buffer_size,
             self._enabled,
             self._seed,
+            self._rng_state,
             self._valid_iterator_id,
             self._number_of_samples_yielded,
-            rng_state,
+            self._snapshot_state,
         ) = state
         self._rng = random.Random()
-        self._rng.setstate(rng_state)
+        self._rng.setstate(self._rng_state)
         self._buffer = []
+        if self._snapshot_state == _SnapshotState.Iterating:
+            self._snapshot_state = _SnapshotState.Restored
```

Comment:
This part is causing `test_simple_snapshot_graph_with_serialization` to fail.

The reason is that `_simple_graph_snapshot_restoration` was written under the assumption that the restoration function will change the `_SnapshotState` from `Iterating` to `Restored` (rather than being done by `__setstate__`). I believe my thinking at the time was that `__setstate__` can be invoked by for reasons unrelated to snapshot restoration (e.g. serialization/multiprocessing), therefore I decided to let the restoration function change the state instead.

Reply:
If you agree that `__setstate__` can be invoked for reasons unrelated for snapshotting, then we should let the restoration function handles the change in state. Let me know what you think.

Reply:
Here is the problem by removing this part of state transition:
As this PR would `reset` for the case of `NotStarted` and `Iterating` for shuffler, after deserialization, the state becomes `NotStarted` no matter serialization happens before the iteration or in the middle of iteration. Then, the DataPipe will always be `reset` once. Then, we would lose the restored `rng_state`.

Reply:
See comment: https://github.com/pytorch/pytorch/pull/83535/files#r951851003

Reply:
This part is removed.